tools/PI/DevHome.PI/Views/BarWindowHorizontal.xaml.cs

```diff
@@ -606,15 +612,89 @@ private void WindowEx_WindowStateChanged(object sender, WindowState e)
-                if (_viewModel.ShowingExpandedContent && Height == FloatingHorizontalBarHeight)
+                Windows.Win32.UI.WindowsAndMessaging.WINDOWPOS wndPos = Marshal.PtrToStructure<Windows.Win32.UI.WindowsAndMessaging.WINDOWPOS>(lParam);
+
+                // We only care about this message if it's triggering a resize
+                if (wndPos.flags.HasFlag(Windows.Win32.UI.WindowsAndMessaging.SET_WINDOW_POS_FLAGS.SWP_NOSIZE))
+                {
+                    break;
+                }
+
+                int floatingBarHeight = CommonHelper.MulDiv(FloatingHorizontalBarHeight, (int)this.GetDpiForWindow(), 96);
+
+                if (PInvoke.IsWindowArranged(hWnd))
+                {
+                    _isSnapped = true;
+                }
+                else
+                {
+                    if (_isSnapped)
+                    {
+                        _transitionFromSnapped = true;
+                        _isSnapped = false;
+                    }
+                }
+
+                if (!_isSnapped && !_viewModel.ShowingExpandedContent && wndPos.cy > floatingBarHeight)
                 {
-                    Height = Settings.Default.ExpandedWindowHeight;
+                    // Enforce our height limit if we're not showing expanded content and we're not being snapped
+                    wndPos.cy = CommonHelper.MulDiv(FloatingHorizontalBarHeight, (int)this.GetDpiForWindow(), 96);
+                    Marshal.StructureToPtr(wndPos, lParam, true);
+                    Debug.WriteLine("WM_WINDOWPOSCHANGING: Enforcing height limit " + _isSnapped + " " + _transitionFromSnapped);
                 }
+                else if (wndPos.cy <= floatingBarHeight && _viewModel.ShowingExpandedContent && _transitionFromSnapped)
+                {
+                    // If we're transitioning from snapped (which always expands our bar) to unsnapped, be sure set our height to the expanded height
+                    wndPos.cy = CommonHelper.MulDiv((int)Settings.Default.ExpandedWindowHeight, (int)this.GetDpiForWindow(), 96);
+                    Marshal.StructureToPtr(wndPos, lParam, true);
+                    Debug.WriteLine("WM_WINDOWPOSCHANGING: Expanding window size for expanded content " + _isSnapped);
+                }
+                else if (wndPos.cy > floatingBarHeight && !_viewModel.ShowingExpandedContent)
+                {
+                    // Our window is bigger than the floating bar height, so we should show the expanded content
+                    _viewModel.ShowingExpandedContent = true;
+                    Debug.WriteLine("WM_WINDOWPOSCHANGING: enabling expanded content due to large window size " + PInvoke.IsWindowArranged(hWnd));
+                }
+                else
+                {
+                    // We'll say we're done transitioning from snapped once our size is "valid" for our current state
+                    _transitionFromSnapped = false;
+                }

```

Comment:
Can this be grouped based on (wndPos.cy > floatingBarHeight) or _viewModel.ShowingExpandedContent for better readability?

Reply:
something like this maybe?
```
if (!_viewModel.ShowingExpandedContent && wndPos.cy > floatingBarHeight)
{
    if (!_isSnapped)
    {
        // Enforce our height limit if we're not showing expanded content and we're not being snapped
        wndPos.cy = CommonHelper.MulDiv(FloatingHorizontalBarHeight, (int)this.GetDpiForWindow(), 96);
        Marshal.StructureToPtr(wndPos, lParam, true);
        Debug.WriteLine("WM_WINDOWPOSCHANGING: Enforcing height limit " + _isSnapped + " " + _transitionFromSnapped);
    }
    else
    {
        // Our window is bigger than the floating bar height, so we should show the expanded content
        _viewModel.ShowingExpandedContent = true;
        Debug.WriteLine("WM_WINDOWPOSCHANGING: enabling expanded content due to large window size " + PInvoke.IsWindowArranged(hWnd));
    }
}
else if (wndPos.cy <= floatingBarHeight && _viewModel.ShowingExpandedContent && _transitionFromSnapped)
{
    // If we're transitioning from snapped (which always expands our bar) to unsnapped, be sure set our height to the expanded height
    wndPos.cy = CommonHelper.MulDiv((int)Settings.Default.ExpandedWindowHeight, (int)this.GetDpiForWindow(), 96);
    Marshal.StructureToPtr(wndPos, lParam, true);
    Debug.WriteLine("WM_WINDOWPOSCHANGING: Expanding window size for expanded content " + _isSnapped);
}
else
{
    // We'll say we're done transitioning from snapped once our size is "valid" for our current state
    _transitionFromSnapped = false;
}
```

Reply:
I like it. Updated.